<a href="https://colab.research.google.com/github/shemaaali/DS-Unit-1-Sprint-2-Statistics/blob/master/module1LS_DS_121_t_tests/S_A5_LS_DS18_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [175]:
### YOUR WORK HERE
import pandas as pd
import numpy as np
from scipy import stats

In [176]:
column_headers = ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'storke', 'compression-ratio', 'horsepower', 'peak-rmp', 'city-mpg', 'highway-mpg', 'price']

In [177]:
auto_mob = 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'
df = pd.read_csv(auto_mob, names=column_headers, na_values='?')
print(df.shape)
df.head()

(205, 26)


symboling  normalized-losses         make  ... city-mpg highway-mpg    price
0          3                NaN  alfa-romero  ...       21          27  13495.0
1          3                NaN  alfa-romero  ...       21          27  16500.0
2          1                NaN  alfa-romero  ...       19          26  16500.0
3          2              164.0         audi  ...       24          30  13950.0
4          2              164.0         audi  ...       18          22  17450.0

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [178]:
### YOUR WORK HERE
# Random State set for reproducibility
sample = df.sample(10, random_state=30)

sample

symboling  normalized-losses  ... highway-mpg    price
121          1              154.0  ...          38   6692.0
176         -1               65.0  ...          32  10898.0
24           1              148.0  ...          38   6229.0
88          -1              137.0  ...          30   9279.0
72           3              142.0  ...          18  35056.0
103          0              108.0  ...          25  13499.0
26           1              148.0  ...          38   7609.0
20           0               81.0  ...          43   6575.0
155          0               91.0  ...          32   8778.0
185          2               94.0  ...          34   8195.0

[10 rows x 26 columns]

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [182]:
### YOUR WORK HERE
sample['highway-mpg'].mean()

32.8

In [183]:
stats.ttest_1samp(sample['highway-mpg'], 35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

Conclusion:

Based on the conclusion of t-test or statistic=-0.95 and p-value=0.36, I reject the null hypothesis that support the salesman (cars sells typically have a fuel efficiency of about 35 miles per gallon on the highway) on the same rate of car sells on the fuel efficiency on the highway, and I will suggest the alterantive that support at the different level of fuel efficiency on the highway.

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [184]:
### YOUR WORK HERE
# Random State set for reproducibility
sample1 = df.sample(100, random_state=30)

sample1

symboling  normalized-losses  ... highway-mpg    price
121          1              154.0  ...          38   6692.0
176         -1               65.0  ...          32  10898.0
24           1              148.0  ...          38   6229.0
88          -1              137.0  ...          30   9279.0
72           3              142.0  ...          18  35056.0
..         ...                ...  ...         ...      ...
189          3                NaN  ...          29  11595.0
192          0                NaN  ...          38  13845.0
171          2              134.0  ...          30  11549.0
203         -1               95.0  ...          27  22470.0
47           0              145.0  ...          19  32250.0

[100 rows x 26 columns]

In [185]:
sample1['highway-mpg'].mean()

30.64

In [186]:
stats.ttest_1samp(sample1['highway-mpg'], 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

Conclusion:

Based on the conclusion of t-test or statistic=-6.518 and p-value=~0, I reject the null hypothesis that support the salesman (cars sells typically have a fuel efficiency of about 35 miles per gallon on the highway) on the same rate of car sells on the fuel efficiency on the highway, and I will suggest the alterantive that support at the different level of fuel efficiency on the highway.

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

Because when we increase the sample size, our t-test or statistic is going to be negative decressed, and we end up with different results for nulll hypothesis and alterantive.

Also, the sample size 1 with 10 cars is much greater than sample size with 100 cars.


## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [40]:
### YOUR WORK HERE

# Grab the file from UCI:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-07-13 20:35:39--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.06s   

2020-07-13 20:35:39 (286 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [41]:
column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

df = pd.read_csv('house-votes-84.data', names=column_headers, na_values="?")

df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n          NaN
2    democrat                 NaN             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [42]:
df = df.replace({'y': 1, 'n': 0})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [43]:
df.describe()

handicapped-infants  water-project  ...   duty-free  south-africa
count           423.000000     387.000000  ...  407.000000    331.000000
mean              0.442080       0.503876  ...    0.427518      0.812689
std               0.497222       0.500632  ...    0.495327      0.390752
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      1.000000
50%               0.000000       1.000000  ...    0.000000      1.000000
75%               1.000000       1.000000  ...    1.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

In [44]:
df.isnull().sum()

party                     0
handicapped-infants      12
water-project            48
budget                   11
physician-fee-freeze     11
el-salvador-aid          15
religious-groups         11
anti-satellite-ban       14
aid-to-contras           15
mx-missile               22
immigration               7
synfuels                 21
education                31
right-to-sue             25
crime                    17
duty-free                28
south-africa            104
dtype: int64

In [50]:
# Compare republicans to democrats(filtering)

rep = df[df['party']=='republican']
dem = df[df['party']=='democrat']

In [48]:
print(rep.shape)
rep.head()

(168, 17)


party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [49]:
print(dem.shape)
dem.head()

(267, 17)


party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [58]:
### YOUR WORK HERE

dem['handicapped-infants'].mean()

0.6046511627906976

In [57]:
rep['handicapped-infants'].mean()

0.18787878787878787

1) Null Hypothesis:

Democrats support this bill at the same level that republicans do.


$H_0: \overline{x}_r == \overline{x}_d$

2) Alternative Hypothesis:

Democrats support this bill in a different proportion than republicans.


$H_a: \overline{x}_r \neq \overline{x}_d$

3) Confidence Level: 99%

In [59]:
#dems_sample = dem['handicapped-infants'].dropna()
#reps_sample = rep['handicapped-infants'].dropna()

In [60]:
stats.ttest_ind(rep['handicapped-infants'], dem['handicapped-infants'], nan_policy='omit')

Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18)


4) Conclusion:

Due to a t-statistic of -9.2 and a p-value of ~0 we reject the null hypothesis that democrats and repulicans support the handicapped-infants bill at the same rate, and suggest the alternative that there are differing levels of support for this bill between the two parties.

The sign of the t-statistic tells us which sample mean is greater.

positive t-statistic: left-hand sample mean is greater

negative t-statistic: right-hand sample mean is greater

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [361]:
### YOUR WORK HERE

a = [1,2,3,4,5,6]
s1 = .6

b =[2, 4, 6, 8, 9, 10]
s2 = .7

coinflips = [a,b] 
def coinflips(coin_samp):
 for x in coin_samp:
  if coin_samp == .6:
    if coin_samp == .7:
     return coin_samp
print(coinflips)

<function coinflips at 0x7f2409cc6730>


In [362]:
import pandas as pd
import numpy as np
from scipy import stats

coinflips = pd.DataFrame({'a': [1,2,3,4,5,6],'b': [2, 4, 6, 8, 9, 10]})
print(coinflips.shape)
coinflips.head()

(6, 2)


a  b
0  1  2
1  2  4
2  3  6
3  4  8
4  5  9

In [363]:
coinflips['a'].mean()

3.5

In [364]:
coinflips['b'].mean()

6.5

In [365]:
stats.ttest_ind(coinflips['a'], coinflips['b'], nan_policy='omit')

Ttest_indResult(statistic=-2.0380986614602725, pvalue=0.0688659774333155)

Conclusion:

Based on a t-statistic of -2.038 and a p-value of 0.06. I fail to reject the null hypothesis(p-value is greater than 0.05) that support a and b at the same rate, and I will suggest the alternative that support a and b at the different rate.



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [410]:
### YOUR WORK HERE

# Using 1 sample

import pandas as pd
import numpy as np
from scipy import stats

coinflips_coin = pd.DataFrame({'a': ['coin', 'non-coin', 'coin', 'non-coin', 'coin', 'non-coin'],'b': [2, 4, 6, 8, 9, 10], 'c': [11,12,13,14,15,16]})
print(coinflips_coin.shape)
coinflips_coin.head()

(6, 3)


a  b   c
0      coin  2  11
1  non-coin  4  12
2      coin  6  13
3  non-coin  8  14
4      coin  9  15

In [411]:
print(coin.shape)
coin.head()

(3, 3)


a  b   c
0  coin  2  11
2  coin  6  13
4  coin  9  15

In [412]:
print(no_coin.shape)
no_coin.head()

(3, 3)


a   b   c
1  non-coin   4  12
3  non-coin   8  14
5  non-coin  10  16

In [413]:
coinflips_coin.mean()

b     6.5
c    13.5
dtype: float64

In [414]:
coin_sample = coinflips_coin['b'].dropna()

In [416]:
coin_sample.mean()

6.5

In [417]:
stats.ttest_1samp(coinflips_coin['b'], 6)

Ttest_1sampResult(statistic=0.3973597071195132, pvalue=0.7074980313783434)


Conclusion:

Based on a t-statistic of 0.39 and a p-value of 0.70. I reject the null hypothesis that is a fair coin, and I will suggest the alternative that is not fair coin.

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [395]:
### YOUR WORK HERE

import pandas as pd
import numpy as np
from scipy import stats

coinflips_coin = pd.DataFrame({'a': ['coin', 'non-coin', 'coin', 'non-coin', 'coin', 'non-coin'],'b': [2, 4, 6, 8, 9, 10], 'c': [11,12,13,14,15,16]})
print(coinflips_coin.shape)
coinflips_coin.head()


(6, 3)


a  b   c
0      coin  2  11
1  non-coin  4  12
2      coin  6  13
3  non-coin  8  14
4      coin  9  15

In [401]:
coin1 = coinflips_coin[coinflips_coin['a']=='coin']
no_coin2 = coinflips_coin[coinflips_coin['a']=='non-coin']

In [406]:
coin1['c'].mean()

5.666666666666667

In [407]:
no_coin2['c'].mean()

14.0

In [408]:
stats.ttest_ind(no_coin2['c'], coin1['c'], nan_policy='omit')

Ttest_indResult(statistic=0.6123724356957945, pvalue=0.5733922538253555)

Conclusion:

Based on a t-statistic of 0.612 and a p-value of 0.573. I reject the null hypothesis that support coin and no coin at the same rate, and I will suggest the alternative that support coin and no coin at the different rate.